### ISSUE TITLE
Allow for the portal nexus to be used on fruit/tree runs too

### ISSUE THREAD

**Issue by othyn** [Speaax/Farming-Helper#5]
On fruit/tree runs, the portal nexus currently isn't a selectable option for;
- Brimhaven (Ardougne)
- Falador
- Lumbridge
- Varrock
- Taverly (Falador)
- etc.

The above teleports like it is for herb runs. Would be nice to have those as an option :)

Also having an option for Camelot being a valid teleport location for Catherby, given the Catherby teleport has higher requirements.

**Misc**
- [x] General UI/UX refresh to show start/stop behaviour

**Herb teleport options to add:**
- [x] Ardougne:
    - (New) Portal Nexus (Ardougne)
- [x] Catherby:
    - (New) Portal Nexus (Camelot)
- [x] Falador:
    - (New) Portal Nexus (Falador)

**Tree teleport options to add:**
- [x] Falador:
    - (New) Portal Nexus (Falador)
- [x] Lumbridge:
    - (New) Portal Nexus (Lumbridge)
- [x] Taverly:
    - (New) Portal Nexus (Falador)
- [x] Varrock:
    - (New) Portal Nexus (Varrock)

**Fruit Tree teleport options to add:**
- [x] Brimhaven:
    - (New) Portal Nexus (Ardougne)
- [x] Catherby:
    - (New) Portal Nexus (Camelot)

**Comment by Speaax** [comment]
I'll take a look at it, many options should be fairly easy to add in since I already have code for it. Fairy ring and spirit trees will require some new logic to be added in and some new configs. So those will probably have to wait.

**Comment by othyn** [comment]
Yeah all those teleport options currently exist for herb runs, so I'm hoping its just a mild refactor. I can look at submitting a PR for this as well, quite happy to look into it.

**Comment by Speaax** [comment]
I had no idea what i was doing when creating this plugins so the code might be a bit messy for you. But to add new teleport options you have to fill in one of these(Found in Herb/Tree/FruitTreeRunItemAndLocation.java) and add in the options in FarmingHelperConfig:
![image](https://github.com/Speaax/Farming-Helper/assets/28015430/0e3fed5e-560c-4dbe-ac1d-dfda95ac6cb7)

I don't mind explaining how it works and how you can fill in each line if you'd like.

**Comment by othyn** [comment]
How do you determine the `interfaceGroupId`, `interfaceChildId` and `regionId`?

**Comment by Speaax** [comment]
![image](https://github.com/Speaax/Farming-Helper/assets/28015430/ba90408c-4820-4570-841c-e772bce3b3aa)

If you launch RuneLite thought IntelliJ and click an interface option it will print out the group/child Id. If the teleport option added in doesn't require an interface to choose a teleport option i set those values to 0.

As for regionId, that's the id of the destination of the teleport. The code uses this information to know the user has teleported and moves on with the code. A website like this can be used to find the regionId : https://explv.github.io/?centreX=2980&centreY=3316&centreZ=0&zoom=7 Just enable region grid and labels.
![image](https://github.com/Speaax/Farming-Helper/assets/28015430/a9745834-9b1f-4d26-bbd9-d4a8e09fce60)

**Comment by othyn** [comment]
Fantastic, thank you for that!

I'll look at adding a load of teleport options tonight 👍 if I can get my environment configured, Graddle gives me nightmares...

**Comment by Speaax** [comment]
Got discord? I could send you some screenshots with the steps i take to set it up if you like.

Discord: .mrbond

**Comment by othyn** [comment]
Got it sorted haha, IntelliJ just read the Gradle file and set everything up. Even had the JDK installed, was far easier than getting things setup for Minecraft mods haha

**Comment by othyn** [comment]
Just to let you know I am working on this, a PR will be along in the next few days.

The first PR is a major refactor of how `*ItemAndLocation` classes are handled. There is now a base `ItemAndLocation` class which is extended from containing a significant amount of overlapping functionality. All location definitions for both `Location` and `WorldPoint` objects are centralised into the base `ItemAndLocation` class. Same goes for Enum option definitions in the `FarmingHelperConfig` class, as those are also currently duplicated per option definition.

All this means that you can easily assign and re-use location sets, just passing in what you are wanting to travel to; Herb Patch, Tree or Fruit Tree.

Once the above is done, it will be incredibly easy to recycle and centrally define new teleports, locations and location sets.

**Comment by Speaax** [comment]
That sounds like a good idea to me. Do let me know if there is other parts
of the code I can help you understand, but it does seem like you have a
better understanding in general :p

Thanks for taking a look at it, I do think a plugin like this would be very
nice if it was in a little better shape.

lør. 4. nov. 2023, 01:26 skrev Ben ***@***.***>:

**Comment by othyn** [comment]
I'll submit a PR for this later, the refactor _kinda_ worked. Some of it is on the PR I'll submit. Still thinking on ways to properly perform a decent refactor. Still getting my head around some of the code.

**Comment by othyn** [comment]
PR #6 submitted for this.

There is currently a bug with the Camelot option not being selected for Catherby teleports, same with Falador for Taverly. This is as on line 903 of FarmingTeleportOverlay, its currently using location.getName() instead of the name of the Teleport. Well, I think thats the issue anyway.

Not sure what the easy solve on that is, without reading the location name from the Teleport and setting it against each Teleport... which isn't great.
